// tslint-misc-rules/rules/camelCaseLocalFunctionsRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint/lib");
var ts = require("typescript");
var nodeIsKind_1 = require("../helpers/nodeIsKind");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithFunction(sourceFile, walk);
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
    ts.forEachChild(ctx.sourceFile, function cb(node) {
        if (nodeIsKind_1.nodeIsKind(node, 'CallExpression') &&
            nodeIsKind_1.nodeIsKind(node.expression, 'Identifier')) {
            checkFunctionName(ctx, node.expression);
        }
        return ts.forEachChild(node, cb);
    });
}
var whitelist = ['Array', 'Boolean', 'Error', 'Function', 'Number', 'Object', 'String'];
function checkFunctionName(ctx, name) {
    var firstLetter = name.text.charAt(0);
    if (firstLetter !== firstLetter.toLowerCase() &&
        whitelist.indexOf(name.text) === -1) {
        ctx.addFailureAtNode(name, 'local function names should be camelCase');
    }
}
