// tslint-misc-rules/rules/classMethodNewlinesRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint/lib");
var ts = require("typescript");
var getClassMethods_1 = require("../helpers/getClassMethods");
var getLeadingWhitespace_1 = require("../helpers/getLeadingWhitespace");
var nodeIsKind_1 = require("../helpers/nodeIsKind");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithFunction(sourceFile, walk);
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
    ts.forEachChild(ctx.sourceFile, function cb(node) {
        if (nodeIsKind_1.nodeIsKind(node, 'ClassDeclaration') || nodeIsKind_1.nodeIsKind(node, 'ClassExpression')) {
            checkClass(ctx, node);
        }
        return ts.forEachChild(node, cb);
    });
}
function checkClass(ctx, node) {
    var sf = ctx.sourceFile;
    var methods = getClassMethods_1.getClassMethods(node);
    methods.reduce(function (previousMethod, method) {
        var leadingWhitespace = getLeadingWhitespace_1.getLeadingWhitespace(method);
        var newlineCount = leadingWhitespace.match(/\n/g).length;
        var hasComments = /\/\/|\/\*\*/g.test(leadingWhitespace);
        var isFirstMethod = method === node.members[0];
        var isInOverloadGroup = method !== previousMethod &&
            method.name.getText(sf) === previousMethod.name.getText(sf);
        var expectedNewlines = isFirstMethod || isInOverloadGroup ? 1 : 2;
        if (newlineCount < expectedNewlines ||
            (newlineCount > expectedNewlines && !hasComments)) {
            var newLine = leadingWhitespace.match('\r\n') ? '\r\n' : '\n';
            ctx.addFailureAtNode(method.name, 'class methods should be preceded by an empty line', Lint.Replacement.appendText(method.getStart(sf) - leadingWhitespace.length, newLine));
        }
        return method;
    }, methods[0]);
}
