// tslint-misc-rules/rules/declareClassMethodsAfterUseRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint/lib");
var ts = require("typescript");
var getClassMethods_1 = require("../helpers/getClassMethods");
var nodeIsKind_1 = require("../helpers/nodeIsKind");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new DeclareClassMethodsAfterUseWalker(sourceFile, this.ruleName, this.getOptions()));
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var DeclareClassMethodsAfterUseWalker = /** @class */ (function (_super) {
    __extends(DeclareClassMethodsAfterUseWalker, _super);
    function DeclareClassMethodsAfterUseWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DeclareClassMethodsAfterUseWalker.prototype.walk = function (sourceFile) {
        var _this = this;
        var cb = function (node) {
            if (nodeIsKind_1.nodeIsKind(node, 'ClassDeclaration') ||
                nodeIsKind_1.nodeIsKind(node, 'ClassExpression')) {
                _this.validate(node);
            }
            return ts.forEachChild(node, cb);
        };
        ts.forEachChild(sourceFile, cb);
    };
    DeclareClassMethodsAfterUseWalker.prototype.validate = function (node) {
        var _this = this;
        this.visitedMethodDeclarations = [];
        this.visitedMethodCalls = [];
        for (var _i = 0, _a = getClassMethods_1.getClassMethods(node); _i < _a.length; _i++) {
            var method = _a[_i];
            this.currentMethodName = method.name.getText(this.getSourceFile());
            this.visitedMethodDeclarations.push(this.currentMethodName);
            ts.forEachChild(method, function (child) {
                _this.visitChildren(child);
            });
        }
    };
    DeclareClassMethodsAfterUseWalker.prototype.visitChildren = function (node) {
        var _this = this;
        ts.forEachChild(node, function (child) {
            if (nodeIsKind_1.nodeIsKind(child, 'CallExpression')) {
                _this.visitCallExpressionInMethod(child);
            }
            _this.visitChildren(child);
        });
    };
    DeclareClassMethodsAfterUseWalker.prototype.visitCallExpressionInMethod = function (node) {
        if (!this.callExpressionBelongsToThis(node.expression)) {
            return;
        }
        var propertyExpression = node.expression;
        var methodName = propertyExpression.name.text;
        if (this.methodHasBeenDeclared(methodName)) {
            if (!this.isRecursion(methodName) && !this.methodHasBeenCalled(methodName)) {
                this.addFailureAtNode(propertyExpression, 'declare class methods after use');
            }
        }
        else {
            // declaration needs to come after first use, not all uses.
            // once we've seen a callsite before a declaration, don't
            // error on any future callsites for that method
            this.visitedMethodCalls.push(methodName);
        }
    };
    DeclareClassMethodsAfterUseWalker.prototype.callExpressionBelongsToThis = function (node) {
        return (nodeIsKind_1.nodeIsKind(node, 'PropertyAccessExpression') &&
            nodeIsKind_1.nodeIsKind(node.expression, 'ThisKeyword'));
    };
    DeclareClassMethodsAfterUseWalker.prototype.methodHasBeenDeclared = function (name) {
        return this.visitedMethodDeclarations.indexOf(name) > -1;
    };
    DeclareClassMethodsAfterUseWalker.prototype.methodHasBeenCalled = function (name) {
        return this.visitedMethodCalls.indexOf(name) > -1;
    };
    DeclareClassMethodsAfterUseWalker.prototype.isRecursion = function (name) {
        return this.currentMethodName === name;
    };
    return DeclareClassMethodsAfterUseWalker;
}(Lint.AbstractWalker));
