// tslint-misc-rules/rules/jsxNoClosingBracketNewlineRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint/lib");
var ts = require("typescript");
var getLeadingWhitespace_1 = require("../helpers/getLeadingWhitespace");
var nodeIsKind_1 = require("../helpers/nodeIsKind");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithFunction(sourceFile, walk);
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
    ts.forEachChild(ctx.sourceFile, function cb(node) {
        if (nodeIsKind_1.nodeIsKind(node, 'JsxElement')) {
            check(ctx, node.openingElement);
        }
        else if (nodeIsKind_1.nodeIsKind(node, 'JsxSelfClosingElement')) {
            check(ctx, node);
        }
        return ts.forEachChild(node, cb);
    });
}
function check(ctx, node) {
    for (var _i = 0, _a = findClosingTokens(node); _i < _a.length; _i++) {
        var token = _a[_i];
        if (getLeadingWhitespace_1.getLeadingWhitespace(token).match(/\n/g)) {
            var fix = Lint.Replacement.deleteText(token.getFullStart(), token.getStart(ctx.sourceFile) - token.getFullStart());
            ctx.addFailureAtNode(token, 'closing brackets for jsx elements should not be on newlines', fix);
        }
    }
}
function findClosingTokens(node) {
    return node
        .getChildren()
        .filter(function (child) {
        return child.kind === ts.SyntaxKind.SlashToken ||
            child.kind === ts.SyntaxKind.GreaterThanToken;
    });
}
