// tslint-misc-rules/rules/noBracesForSingleLineArrowFunctionsRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint/lib");
var ts = require("typescript");
var nodeIsKind_1 = require("../helpers/nodeIsKind");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new Walker(sourceFile, this.ruleName, this.getOptions()));
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var Walker = /** @class */ (function (_super) {
    __extends(Walker, _super);
    function Walker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Walker.prototype.walk = function (sourceFile) {
        var _this = this;
        var cb = function (node) {
            if (nodeIsKind_1.nodeIsKind(node, 'ArrowFunction')) {
                _this.validate(node);
            }
            return ts.forEachChild(node, cb);
        };
        ts.forEachChild(sourceFile, cb);
    };
    Walker.prototype.validate = function (node) {
        var body = node.body;
        if (this.functionBodyIsBraced(body) &&
            this.functionBodyHasOneStatement(body) &&
            this.functionBodyIsOneLine(body)) {
            this.addFailureAtNode(body, 'single-line arrow functions should not be wrapped in braces', Lint.Replacement.replaceNode(body, this.getFixedText(body)));
        }
    };
    Walker.prototype.functionBodyIsBraced = function (node) {
        return nodeIsKind_1.nodeIsKind(node, 'Block');
    };
    Walker.prototype.functionBodyHasOneStatement = function (node) {
        return node.statements.length === 1;
    };
    Walker.prototype.functionBodyIsOneLine = function (node) {
        var bodyText = node.getFullText(this.getSourceFile());
        return !/\n/.test(bodyText);
    };
    Walker.prototype.getFixedText = function (node) {
        var sf = this.getSourceFile();
        var body = node.getChildAt(1, sf);
        var result = this.stripSemicolon(body.getText(sf));
        var statement = body.getChildAt(0, sf);
        if (nodeIsKind_1.nodeIsKind(statement, 'ReturnStatement')) {
            result = result.replace('return', '').trim();
            var returnExpression = statement.getChildAt(1, sf);
            if (nodeIsKind_1.nodeIsKind(returnExpression, 'ObjectLiteralExpression')) {
                result = "(" + result + ")";
            }
        }
        return result;
    };
    Walker.prototype.stripSemicolon = function (text) {
        return text.trim().replace(/;$/, '');
    };
    return Walker;
}(Lint.AbstractWalker));
