// tslint-misc-rules/rules/noPropertyInitializersRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint/lib");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new ClassMethodNewlinesWalker(sourceFile, this.getOptions()));
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var ClassMethodNewlinesWalker = /** @class */ (function (_super) {
    __extends(ClassMethodNewlinesWalker, _super);
    function ClassMethodNewlinesWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ClassMethodNewlinesWalker.prototype.visitClassDeclaration = function (node) {
        _super.prototype.visitClassDeclaration.call(this, node);
        this.validate(node);
    };
    ClassMethodNewlinesWalker.prototype.visitClassExpression = function (node) {
        _super.prototype.visitClassExpression.call(this, node);
        this.validate(node);
    };
    ClassMethodNewlinesWalker.prototype.validate = function (node) {
        for (var _i = 0, _a = node.members; _i < _a.length; _i++) {
            var member = _a[_i];
            if (member.initializer) {
                this.addFailure(this.createFailure(member.name.getStart(), member.name.getWidth(), 'property initializers are nonstandard -- assign in constructor or method'));
            }
        }
    };
    return ClassMethodNewlinesWalker;
}(Lint.RuleWalker));
