// tslint-misc-rules/rules/preferOrOperatorOverTernaryRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint/lib");
var ts = require("typescript");
var nodeIsKind_1 = require("../helpers/nodeIsKind");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithFunction(sourceFile, walk);
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
    ts.forEachChild(ctx.sourceFile, function cb(node) {
        if (nodeIsKind_1.nodeIsKind(node, 'ConditionalExpression')) {
            check(ctx, node);
        }
        return ts.forEachChild(node, cb);
    });
}
function check(ctx, node) {
    var sf = ctx.sourceFile;
    var condition = node.condition, whenTrue = node.whenTrue, questionToken = node.questionToken, colonToken = node.colonToken;
    if (condition.getText(sf) === whenTrue.getText(sf) &&
        ((isIdentifier(condition) && isIdentifier(whenTrue)) ||
            (isPropertyAccessChain(condition) && isPropertyAccessChain(whenTrue)))) {
        var fix = Lint.Replacement.replaceFromTo(questionToken.getStart(sf), colonToken.getStart(sf) + 1, '||');
        ctx.addFailureAtNode(whenTrue, "use '||' when first and second operands of ternary are identical", fix);
    }
}
function isPropertyAccessChain(startNode) {
    if (!isPropertyAccess(startNode) || !isIdentifier(startNode.name)) {
        return false;
    }
    var node = startNode;
    while (isPropertyAccess(node.expression)) {
        node = node.expression;
    }
    return (node.expression &&
        (isIdentifier(node.expression) ||
            node.expression.kind === ts.SyntaxKind.ThisKeyword));
}
function isIdentifier(node) {
    return node && node.kind === ts.SyntaxKind.Identifier;
}
function isPropertyAccess(node) {
    return node && node.kind === ts.SyntaxKind.PropertyAccessExpression;
}
