// tslint-misc-rules/rules/reactLifecycleOrderRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint/lib");
var getClassMethods_1 = require("../helpers/getClassMethods");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new ReactLifecyleOrderRule(sourceFile, this.getOptions()));
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var defaultOrder = [
    'componentWillMount',
    'render',
    'componentDidMount',
    'componentWillReceiveProps',
    'shouldComponentUpdate',
    'componentWillUpdate',
    'componentDidUpdate',
    'componentWillUnmount',
];
var ReactLifecyleOrderRule = /** @class */ (function (_super) {
    __extends(ReactLifecyleOrderRule, _super);
    function ReactLifecyleOrderRule(sourceFile, options) {
        var _this = _super.call(this, sourceFile, options) || this;
        var orderOptions = _this.getOptions();
        _this.expectedOrder =
            orderOptions && orderOptions.length > 0 ? orderOptions : defaultOrder;
        return _this;
    }
    ReactLifecyleOrderRule.prototype.visitClassDeclaration = function (node) {
        this.validate(node);
        _super.prototype.visitClassDeclaration.call(this, node);
    };
    ReactLifecyleOrderRule.prototype.visitClassExpression = function (node) {
        this.validate(node);
        _super.prototype.visitClassExpression.call(this, node);
    };
    ReactLifecyleOrderRule.prototype.validate = function (node) {
        var _this = this;
        if (!this.isReactComponent(node)) {
            return;
        }
        var sf = this.getSourceFile();
        var relevantMethods = getClassMethods_1.getClassMethods(node).filter(function (method) { return _this.expectedOrder.indexOf(method.name.getText(sf)) > -1; });
        var sortedMethods = relevantMethods.slice().sort(function (left, right) {
            var leftName = left.name.getText(sf);
            var rightName = right.name.getText(sf);
            var leftIndex = _this.expectedOrder.indexOf(leftName);
            var rightIndex = _this.expectedOrder.indexOf(rightName);
            return leftIndex > rightIndex ? 1 : -1;
        });
        relevantMethods.forEach(function (method, index) {
            if (sortedMethods[index] !== method) {
                _this.addFailure(_this.createFailure(method.name.getStart(sf), method.name.getWidth(sf), "expected React lifecyle method '" + sortedMethods[index].name.getText(sf) + "'"));
            }
        });
    };
    ReactLifecyleOrderRule.prototype.isReactComponent = function (node) {
        if (!node.heritageClauses || node.heritageClauses.length !== 1) {
            return false;
        }
        var ancestor = node.heritageClauses[0].types[0];
        if (!ancestor) {
            return false;
        }
        return ([
            'React.Component',
            'React.PureComponent',
            'Component',
            'PureComponent',
        ].indexOf(ancestor.getText(this.getSourceFile()).replace(/<.+?>$/, '')) > -1);
    };
    return ReactLifecyleOrderRule;
}(Lint.RuleWalker));
