// tslint-misc-rules/rules/sortImportsRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint/lib");
var ts = require("typescript");
var detect_newline_1 = require("detect-newline");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        var options = {
            whitespaceInsensitive: this.getOptions().ruleArguments.indexOf('whitespace-insensitive') > -1,
        };
        return this.applyWithWalker(new Walker(sourceFile, this.ruleName, options));
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var Walker = /** @class */ (function (_super) {
    __extends(Walker, _super);
    function Walker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Walker.prototype.walk = function (sf) {
        var _this = this;
        var _loop_1 = function (importGroup) {
            var sortableLinesToImports = new Map();
            var unsortedLines = importGroup.map(function (importDeclaration) {
                // opening brace is below alphanumeric characters char-code-wise, but want named imports above defaults
                // + comes directly after *, so swap with that
                var sortableLine = importDeclaration
                    .getText(sf)
                    .toLowerCase()
                    .replace('import {', 'import +');
                if (_this.options.whitespaceInsensitive) {
                    sortableLine = _this.normalizeAllWhitespace(sortableLine);
                }
                sortableLinesToImports.set(sortableLine, importDeclaration);
                return sortableLine;
            });
            var sortedLines = unsortedLines.slice().sort(function (left, right) {
                if (_this.isSideEffectImport(sortableLinesToImports.get(left)) &&
                    _this.isSideEffectImport(sortableLinesToImports.get(right))) {
                    return 0;
                }
                return left > right ? 1 : left < right ? -1 : 0;
            });
            var _loop_2 = function (i) {
                if (unsortedLines[i] !== sortedLines[i] &&
                    !this_1.isSideEffectImport(sortableLinesToImports.get(unsortedLines[i])) &&
                    !this_1.isSideEffectImport(sortableLinesToImports.get(sortedLines[i]))) {
                    var expectedImportIndex = this_1.findIndex(unsortedLines, function (line) { return line === sortedLines[i]; });
                    var expectedImport = importGroup[expectedImportIndex];
                    var actualImport = importGroup[i];
                    var message = this_1.normalizeAllWhitespace(this_1.getFailureMessage(expectedImport, actualImport));
                    var sortedImports = sortedLines.map(function (line) {
                        return sortableLinesToImports
                            .get(line)
                            .getFullText(sf)
                            .trim();
                    });
                    var groupStart = importGroup[0].getStart(sf);
                    var groupEnd = importGroup[importGroup.length - 1].getEnd();
                    var newline = detect_newline_1.graceful(sf.getFullText());
                    var fix = Lint.Replacement.replaceFromTo(groupStart, groupEnd, sortedImports.join(newline));
                    // work around some kind of multiline error span bug in tslint rule testing
                    var failureNode = /\n/.test(actualImport.getText(sf).trim())
                        ? actualImport.getFirstToken()
                        : actualImport;
                    this_1.addFailureAtNode(failureNode, message, fix);
                    return "break";
                }
            };
            for (var i = 0; i < unsortedLines.length; i += 1) {
                var state_1 = _loop_2(i);
                if (state_1 === "break")
                    break;
            }
        };
        var this_1 = this;
        for (var _i = 0, _a = this.getImportGroups(sf); _i < _a.length; _i++) {
            var importGroup = _a[_i];
            _loop_1(importGroup);
        }
    };
    Walker.prototype.getImportGroups = function (sourceFile) {
        var breakGroup = true;
        var importGroups = [[]];
        for (var _i = 0, _a = sourceFile.statements; _i < _a.length; _i++) {
            var statement = _a[_i];
            if (this.isImportStatement(statement)) {
                importGroups[importGroups.length - 1].push(statement);
                breakGroup = true;
            }
            else if (breakGroup) {
                importGroups.push([]);
                breakGroup = false;
            }
        }
        return importGroups.filter(function (group) { return group.length > 0; });
    };
    Walker.prototype.findIndex = function (array, predicate) {
        for (var i = 0; i < array.length; i += 1) {
            if (predicate(array[i])) {
                return i;
            }
        }
    };
    Walker.prototype.isImportStatement = function (node) {
        return (node.kind === ts.SyntaxKind.ImportDeclaration ||
            this.isImportRequireStatement(node));
    };
    Walker.prototype.getFailureMessage = function (expectedImport, actualImport) {
        var expected = this.getImportBindingName(expectedImport);
        var actual = this.getImportBindingName(actualImport);
        return "out-of-order imports: expected '" + expected + "' but saw '" + actual + "'";
    };
    Walker.prototype.getImportBindingName = function (node) {
        var sf = this.getSourceFile();
        if (this.isImportRequireStatement(node)) {
            return node.name.getText(sf);
        }
        else {
            if (node.importClause) {
                return node.importClause.getText(sf);
            }
            return node.moduleSpecifier.getText(sf);
        }
    };
    Walker.prototype.isImportRequireStatement = function (node) {
        return node.kind === ts.SyntaxKind.ImportEqualsDeclaration;
    };
    Walker.prototype.isSideEffectImport = function (node) {
        return !this.isImportRequireStatement(node) && !node.importClause;
    };
    Walker.prototype.normalizeAllWhitespace = function (content) {
        return content.replace(/[\s]+/g, ' ').replace(', }', ' }');
    };
    return Walker;
}(Lint.AbstractWalker));
